src/data/mv_polynomial/equiv.lean

```diff
@@ -319,6 +321,265 @@ begin
+                  polynomial.mul_coeff_zero, finsupp.mem_support_iff, ne.def, eval₂_hom_X',
+                  mul_zero, coeff_zero ],
+      have t : ¬¬(finsupp.cons 0 m) 0 = 0 := by simp only [ not_not, finsupp.cons_zero],
+      simp only [t, if_false] },
+    { rw fin_succ_equiv_X_eq_zero,
+      let i' := nat.pred i,
+      have r : i = i'.succ,
+      { rw nat.succ_pred_eq_of_pos _,
+        exact nat.pos_of_ne_zero (by simpa) },
+      rw [r, polynomial.coeff_mul_X, hp i'],
+      simp only [finsupp.mem_support_iff, ne.def, finsupp.cons_zero],
+      rw ← r,
+      simp only [c_i, if_true, not_false_iff],
+      congr,
+      ext,
+      by_cases c_a : a = 0,
+      { rw c_a,
+        simp only [finsupp.single_eq_same, finsupp.coe_tsub, pi.sub_apply],
+        repeat { rw finsupp.cons_zero },
+        refl },
+      { simp only [finsupp.single_eq_same, finsupp.coe_tsub, pi.sub_apply],
+        have c_a' : a ≠ 0 := by simpa using c_a,
+        rw [←fin.succ_pred a c_a', finsupp.cons_succ, finsupp.cons_succ, finsupp.single],
+        have c_a'' : ¬ 0 = a := by simpa using c_a'.symm,
+        simp [c_a'', if_false] } } },
+  { let j' := fin.pred j c_j,
+    have r : j = j'.succ := by simp,
+    rw [r, fin_succ_equiv_X_ne_zero, polynomial.coeff_mul_C, coeff_mul_X', hp i],
+    simp only [fin.succ_pred, finsupp.mem_support_iff, ne.def],
+    by_cases c_mj' : m j' = 0,
+    { simp only [r, finsupp.cons_succ, c_mj', if_false, eq_self_iff_true, not_true] },
+    { simp only [r, finsupp.cons_succ, c_mj', if_true, not_false_iff],
+      congr,
+      ext,
+      by_cases c_a : a = 0,
+      { simp [c_a, finsupp.cons_zero, finsupp.single, c_j] },
+      { simp only [finsupp.coe_tsub, pi.sub_apply],
+        rw ←fin.succ_pred a c_a,
+        repeat {rw finsupp.cons_succ},
+        simp [finsupp.single, c_j],
+        congr } } },
+end
+
+/-- This gives the relation between the coefficients of the coefficients of `fin_succ_equiv R n f`
+    and the coefficients of `f`. -/
+lemma fin_succ_equiv_coeff_coeff {n : ℕ} {R : Type u} [comm_semiring R]
+  (m : fin n →₀  ℕ) (f : mv_polynomial (fin (n + 1)) R ) (i : ℕ) :
+  coeff m (polynomial.coeff (fin_succ_equiv R n f) i) = coeff (finsupp.cons i m) f :=
+begin
+  revert i m,
+  apply induction_on f,
+  apply fin_succ_equiv_coeff_coeff_C,
+  intros p q hp hq i m,
+  simp only [(fin_succ_equiv R n).map_add, polynomial.coeff_add, coeff_add, hp, hq],
+  apply fin_succ_equiv_coeff_coeff_case_p_X,
+end
+
+lemma eval_eq_eval_mv_eval' {n : ℕ} {R : Type u} [comm_semiring R] (s : fin n → R) (y : R)
+  (f : mv_polynomial (fin (n + 1)) R) : eval (fin.cons y s : fin (n + 1) → R) f
+  = polynomial.eval y (polynomial.map (eval s) ((fin_succ_equiv R n) f)) :=
+begin
+  apply induction_on f,
+  intro a,
+  simp,
+  intros p q hp hq,
+  simp only [(fin_succ_equiv R n).map_add, ring_hom.map_add, polynomial.map_add,
+             polynomial.eval_add],
+  congr,
+  exact hp,
+  exact hq,
```

Comment:
This suggests to me that there are multiple open goals in the tactic state. Please structure your proofs with `{ .. }` to clearly distinguish these subproofs.

Reply:
Done! I've wrapped each goal with `{ }` in all of my proofs.